lib/codegen/src/nan_canonicalization.rs

```diff
@@ -0,0 +1,120 @@
+//! A NaN-canonicalizing rewriting pass. For instructions that can potentially
+//! result in a nondeterministic NaN value, insert operations to check for NaN,
+//! and replace the result with a deterministic canonical NaN value if the
+//! result of an instruction was in fact a NaN.
+
+use cursor::{Cursor, FuncCursor};
+use ir::{DataFlowGraph, Function, Inst, InstBuilder, InstructionData, Opcode, Value};
+use ir::condcodes::FloatCC;
+use ir::immediates::{Ieee32, Ieee64};
+use ir::types;
+use ir::types::Type;
+use timing;
+
+// Canonical 32-bit and 64-bit NaN values.
+static CANON_32BIT_NAN: u32 = 0b01111111100000000000000000000001;
+static CANON_64BIT_NAN: u64 =
+    0b0111111111110000000000000000000000000000000000000000000000000001;
```

Comment:
We need to set the most significant bit of the significand field to 1, so that these are quiet NaNs rather than signaling NaNs. (For completeness, I'm ignoring pre-NAN2008 MIPS here.) Even though IEEE 754 doesn't say much here, it does say that a NaN returned from arithmetic is a quiet NaN.

It would also be good to set the least significant bit to 0, so that we're consistent with common hardware implementations such as [ARM's "default NaN" mode](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.ddi0274h/Babffifj.html).